File: bot.js
/**
 * Katymio Dumper – Discord bot that dumps Lua / Luau scripts (Node.js).
 *
 * Command:  .l [attachment | url]
 *
 * Accepted input extensions: .lua  .luau  .txt
 * Output:   <original_name>.lua.txt
 * Response: "file successfully dumped in <ms> ms"
 *
 * Security: uploaded files are scanned for dangerous patterns (path-discovery,
 * shell execution, sensitive-file access, heavy obfuscation) both before and
 * after the Lua dumper runs.  Dangerous files are blocked and a Discord embed
 * alarm is sent in the channel.
 */

"use strict";

require("dotenv").config();

const fs          = require("fs");
const os          = require("os");
const path        = require("path");
const { execFileSync, spawnSync } = require("child_process");

const { Client, GatewayIntentBits, EmbedBuilder, AttachmentBuilder, Colors } = require("discord.js");
const axios = require("axios");

const { scanFile } = require("./scanner");

// --------------------------------------------------------------------------- //
// Configuration
// --------------------------------------------------------------------------- //

const TOKEN              = process.env.DISCORD_TOKEN;
const PREFIX             = ".";
const ALLOWED_EXTENSIONS = new Set([".lua", ".luau", ".txt"]);
const DUMPER_SCRIPT      = path.join(__dirname, "dumper.lua");
const MAX_FILE_SIZE      = 8 * 1024 * 1024; // 8 MB – Discord upload limit
const DUMPER_TIMEOUT_MS  = 30_000;           // 30 seconds
const HTTP_TIMEOUT_MS    = 15_000;           // 15 seconds

// --------------------------------------------------------------------------- //
// Helpers
// --------------------------------------------------------------------------- //

/**
 * Return a sanitised file stem (no path separators, limited to 64 chars).
 * @param {string} name
 * @returns {string}
 */
function safeStem(name) {
  const ext  = path.extname(name);
  let   stem = ext ? path.basename(name, ext) : path.basename(name);
  if (stem.startsWith(".")) stem = stem.slice(1);
  stem = stem.replace(/[^\w\-.]/g, "_");
  return (stem.slice(0, 64) || "script");
}

/**
 * Find the first available Lua interpreter on PATH, or null.
 * @returns {string|null}
 */
function findLua() {
  for (const candidate of ["lua", "lua5.4", "lua5.3", "lua5.2", "lua5.1", "luajit"]) {
    const result = spawnSync("which", [candidate], { encoding: "utf8" });
    if (result.status === 0) return candidate;
  }
  return null;
}

const LUA_BIN = findLua();

/**
 * Run the Lua dumper on inputPath, writing the result to outputPath.
 * Returns { success: boolean, error: string }.
 * @param {string} inputPath
 * @param {string} outputPath
 * @returns {{ success: boolean, error: string }}
 */
function runDumper(inputPath, outputPath) {
  if (!LUA_BIN) {
    return { success: false, error: "`lua` interpreter not found on the server" };
  }
  try {
    execFileSync(LUA_BIN, [DUMPER_SCRIPT, inputPath, outputPath], {
      timeout: DUMPER_TIMEOUT_MS,
      encoding: "utf8",
    });
    return { success: true, error: "" };
  } catch (err) {
    if (err.code === "ETIMEDOUT") {
      return { success: false, error: `Dumper timed out (> ${DUMPER_TIMEOUT_MS / 1000} s)` };
    }
    if (err.code === "ENOENT") {
      return { success: false, error: "`lua` interpreter not found on the server" };
    }
    const stderr = (err.stderr || err.stdout || "").toString().trim();
    return { success: false, error: stderr || "Dumper exited with a non-zero code" };
  }
}

// --------------------------------------------------------------------------- //
// Security alert helpers
// --------------------------------------------------------------------------- //

const SEVERITY_EMOJI = { CRITICAL: "🚨", HIGH: "⛔", MEDIUM: "⚠️" };
const SEVERITY_COLOR = {
  CRITICAL: Colors.DarkRed,
  HIGH:     Colors.Red,
  MEDIUM:   Colors.Orange,
};

/**
 * Send a Discord embed alarm when a dangerous file is detected.
 * @param {import('discord.js').Message} message
 * @param {string} filename
 * @param {{ isDangerous: boolean, findings: Array, highestSeverity: string }} result
 * @param {"pre-dump"|"post-dump"} stage
 */
async function alertDangerousFile(message, filename, result, stage = "pre-dump") {
  const top = result.highestSeverity;

  console.warn(
    `[SECURITY ALERT] Dangerous file blocked | stage=${stage} | ` +
    `user=${message.author.tag} (id=${message.author.id}) | ` +
    `channel=#${message.channel.name} (id=${message.channel.id}) | ` +
    `file=${filename} | severity=${top} | ` +
    `findings=${JSON.stringify(result.findings.map((f) => f.name))}`
  );

  const escapedName = filename.replace(/[\\\_*~`|]/g, "\\$&");
  const description =
    stage === "post-dump"
      ? `The file **${escapedName}** was deobfuscated and the resulting code ` +
        "contains dangerous patterns. The output has been withheld."
      : `The file **${escapedName}** was blocked before execution because it ` +
        "contains patterns associated with bot path discovery or remote code execution.";

  const embed = new EmbedBuilder()
    .setTitle(`${SEVERITY_EMOJI[top] ?? "⚠️"} Dangerous File Blocked`)
    .setDescription(description)
    .setColor(SEVERITY_COLOR[top] ?? Colors.Orange)
    .addFields(
      { name: "Submitted by", value: `<@${message.author.id}>`, inline: true },
      { name: "Channel",      value: `<#${message.channel.id}>`, inline: true },
      { name: "Severity",     value: top,                         inline: true },
      { name: "Stage",        value: stage,                       inline: true }
    );

  const findingsLines = result.findings.map(
    (f) => `• **${f.name}** \`[${f.severity}]\` – ${f.description}`
  );
  if (findingsLines.length > 0) {
    embed.addFields({
      name: "Findings",
      value: findingsLines.join("\n").slice(0, 1024),
      inline: false,
    });
  }

  embed.setFooter({ text: "This incident has been logged. Contact an admin if needed." });

  await message.channel.send({ embeds: [embed] });
}

// --------------------------------------------------------------------------- //
// Discord client
// --------------------------------------------------------------------------- //

const client = new Client({
  intents: [
    GatewayIntentBits.Guilds,
    GatewayIntentBits.GuildMessages,
    GatewayIntentBits.MessageContent,
  ],
});

client.once("ready", () => {
  console.log(`Logged in as ${client.user.tag} (id: ${client.user.id})`);
  console.log("Ready. Prefix:", PREFIX);
});

// --------------------------------------------------------------------------- //
// Command handler
// --------------------------------------------------------------------------- //

client.on("messageCreate", async (message) => {
  if (message.author.bot) return;
  if (!message.content.startsWith(`${PREFIX}l`)) return;

  // Allow ".l" or ".l <url>" – strip the command prefix
  const rest = message.content.slice(2).trim(); // everything after ".l"

  let rawName    = null;
  let fileBuffer = null;

  // ── 1. Determine source ──────────────────────────────────────────────────
  if (message.attachments.size > 0) {
    const attachment = message.attachments.first();
    rawName = attachment.name;

    if (attachment.size > MAX_FILE_SIZE) {
      await message.channel.send("❌ File is too large (> 8 MB).");
      return;
    }

    try {
      const resp = await axios.get(attachment.url, {
        responseType: "arraybuffer",
        timeout: HTTP_TIMEOUT_MS,
      });
      fileBuffer = Buffer.from(resp.data);
    } catch (err) {
      await message.channel.send(`❌ Failed to download the attachment: ${err.message}`);
      return;
    }
  } else if (rest) {
    const urlMatch = rest.match(/https?:\/\/\S+/);
    if (!urlMatch) {
      await message.channel.send("❌ Please provide a valid URL or attach a file.");
      return;
    }
    const url = urlMatch[0].replace(/\)$/, "");
    rawName = url.split("/").pop().split("?")[0] || "script.lua";

    try {
      const resp = await axios.get(url, {
        responseType: "arraybuffer",
        timeout: HTTP_TIMEOUT_MS,
        maxContentLength: MAX_FILE_SIZE,
      });
      const data = Buffer.from(resp.data);
      if (data.length > MAX_FILE_SIZE) {
        await message.channel.send("❌ Remote file is too large (> 8 MB).");
        return;
      }
      fileBuffer = data;
    } catch (err) {
      const httpStatus = err.response ? err.response.status : null;
      const msg = httpStatus
        ? `HTTP ${httpStatus} when fetching URL`
        : err.message;
      await message.channel.send(`❌ Failed to download the file: ${msg}`);
      return;
    }
  } else {
    await message.channel.send(
      "❌ Usage: `.l [url]` or attach a `.lua` / `.luau` / `.txt` file."
    );
    return;
  }

  // ── 2. Validate extension ────────────────────────────────────────────────
  const ext = path.extname(rawName).toLowerCase();
  if (!ALLOWED_EXTENSIONS.has(ext)) {
    await message.channel.send(
      `❌ Unsupported file type \`${ext}\`. ` +
      `Accepted: ${[...ALLOWED_EXTENSIONS].sort().join(", ")}`
    );
    return;
  }

  // ── 3.5 Pre-dump security scan ───────────────────────────────────────────
  const preScan = scanFile(fileBuffer);
  if (preScan.isDangerous) {
    await alertDangerousFile(message, rawName, preScan, "pre-dump");
    return;
  }
  if (preScan.findings.length > 0) {
    console.warn(
      `[SECURITY WARNING] Suspicious file (MEDIUM heuristics only) | ` +
      `user=${message.author.tag} (id=${message.author.id}) | ` +
      `file=${rawName} | ` +
      `findings=${JSON.stringify(preScan.findings.map((f) => f.name))}`
    );
  }

  // ── 4. Write to temp dir, run dumper, collect output ─────────────────────
  const stem       = safeStem(rawName);
  const outputName = `${stem}.lua.txt`;

  const tmpDir     = fs.mkdtempSync(path.join(os.tmpdir(), "katymio-"));
  const inputPath  = path.join(tmpDir, `input${ext}`);
  const outputPath = path.join(tmpDir, outputName);

  try {
    fs.writeFileSync(inputPath, fileBuffer);

    const start   = Date.now();
    const { success, error } = runDumper(inputPath, outputPath);
    const elapsedMs = Date.now() - start;

    if (!success) {
      await message.channel.send(`❌ Dumper error: ${error}`);
      return;
    }

    if (!fs.existsSync(outputPath) || fs.statSync(outputPath).size === 0) {
      await message.channel.send("❌ Dumper produced no output.");
      return;
    }

    // ── 4.5 Post-dump security scan (deobfuscated output) ──────────────────
    const postScan = scanFile(fs.readFileSync(outputPath));
    if (postScan.isDangerous) {
      await alertDangerousFile(message, rawName, postScan, "post-dump");
      return;
    }

    const attachment = new AttachmentBuilder(outputPath, { name: outputName });
    await message.channel.send({
      content: `✅ file successfully dumped in ${elapsedMs} ms`,
      files: [attachment],
    });
  } finally {
    // Clean up temp directory
    try { fs.rmSync(tmpDir, { recursive: true, force: true }); } catch { /* ignore */ }
  }
});

// --------------------------------------------------------------------------- //
// Entry point
// --------------------------------------------------------------------------- //

if (!TOKEN) {
  console.error(
    "DISCORD_TOKEN environment variable is not set.\n" +
    "Copy .env.example to .env and fill in your token."
  );
  process.exit(1);
}

client.login(TOKEN);

File: bot.py
"""
Katymio Dumper – Discord bot that dumps Lua / Luau scripts.

Command:  .l [attachment | url]

Accepted input extensions: .lua  .luau  .txt
Output:   <original_name>.lua.txt
Response: "file successfully dumped in <ms> ms"
"""

import asyncio
import os
import re
import subprocess
import tempfile
import time
from pathlib import Path

import aiohttp
import discord
from discord.ext import commands
from dotenv import load_dotenv

load_dotenv()

TOKEN = os.getenv("DISCORD_TOKEN")
PREFIX = "."
ALLOWED_EXTENSIONS = {".lua", ".luau", ".txt"}
DUMPER_SCRIPT = Path(__file__).parent / "dumper.lua"
MAX_FILE_SIZE = 8 * 1024 * 1024  # 8 MB – Discord upload limit

# --------------------------------------------------------------------------- #
# Bot setup
# --------------------------------------------------------------------------- #

intents = discord.Intents.default()
intents.message_content = True

bot = commands.Bot(command_prefix=PREFIX, intents=intents)


# --------------------------------------------------------------------------- #
# Helpers
# --------------------------------------------------------------------------- #

def _safe_stem(name: str) -> str:
    """Return a sanitised file stem (no path separators, limit length)."""
    p = Path(name)
    # Path(".lua").stem == ".lua" (no suffix), Path("foo.lua").stem == "foo"
    stem = p.stem if p.suffix else p.name
    # If the stem is just the extension dot-prefix (e.g. ".lua"), strip it
    if stem.startswith("."):
        stem = stem[1:]
    stem = re.sub(r"[^\w\-.]", "_", stem)
    return stem[:64] or "script"


async def _download_url(session: aiohttp.ClientSession, url: str) -> bytes:
    """Download content from *url*, raising ValueError on problems."""
    async with session.get(url, timeout=aiohttp.ClientTimeout(total=15)) as resp:
        if resp.status != 200:
            raise ValueError(f"HTTP {resp.status} when fetching URL")
        content_length = resp.headers.get("Content-Length")
        if content_length and int(content_length) > MAX_FILE_SIZE:
            raise ValueError("Remote file is too large (> 8 MB)")
        data = await resp.read()
    if len(data) > MAX_FILE_SIZE:
        raise ValueError("Remote file is too large (> 8 MB)")
    return data


def _find_lua() -> str | None:
    """Return the name of the first available Lua interpreter, or None."""
    for candidate in ("lua", "lua5.4", "lua5.3", "lua5.2", "lua5.1", "luajit"):
        if subprocess.run(
            ["which", candidate], capture_output=True
        ).returncode == 0:
            return candidate
    return None


_LUA_BIN: str | None = _find_lua()


def _run_dumper(input_path: Path, output_path: Path) -> tuple[bool, str]:
    """
    Run the Lua dumper on *input_path*, writing result to *output_path*.

    Returns (success, error_message).
    The dumper is invoked as:
        lua dumper.lua <input> <output>
    """
    if _LUA_BIN is None:
        return False, "`lua` interpreter not found on the server"
    try:
        result = subprocess.run(
            [_LUA_BIN, str(DUMPER_SCRIPT), str(input_path), str(output_path)],
            capture_output=True,
            text=True,
            timeout=30,
        )
        if result.returncode != 0:
            stderr = result.stderr.strip() or result.stdout.strip()
            return False, stderr or "Dumper exited with a non-zero code"
        return True, ""
    except FileNotFoundError:
        return False, "`lua` interpreter not found on the server"
    except subprocess.TimeoutExpired:
        return False, "Dumper timed out (> 30 s)"


# --------------------------------------------------------------------------- #
# Command
# --------------------------------------------------------------------------- #

@bot.command(name="l")
async def dump_command(ctx: commands.Context, url: str = ""):
    """
    `.l [url]`  or  `.l` with a file attachment.

    Dumps the provided Lua / Luau / txt script and returns a .lua.txt file.
    """
    attachment = None
    raw_name = None
    file_bytes = None

    # ── 1. Determine source ──────────────────────────────────────────────────
    if ctx.message.attachments:
        attachment = ctx.message.attachments[0]
        raw_name = attachment.filename
    elif url:
        # Accept a bare URL or a markdown-style link [text](url)
        match = re.search(r"https?://\S+", url)
        if not match:
            await ctx.send("❌ Please provide a valid URL or attach a file.")
            return
        url = match.group(0).rstrip(")")
        raw_name = url.split("/")[-1].split("?")[0] or "script.lua"
    else:
        await ctx.send(
            "❌ Usage: `.l [url]` or attach a `.lua` / `.luau` / `.txt` file."
        )
        return

    # ── 2. Validate extension ────────────────────────────────────────────────
    ext = Path(raw_name).suffix.lower()
    if ext not in ALLOWED_EXTENSIONS:
        await ctx.send(
            f"❌ Unsupported file type `{ext}`. "
            f"Accepted: {', '.join(sorted(ALLOWED_EXTENSIONS))}"
        )
        return

    # ── 3. Download content ──────────────────────────────────────────────────
    async with ctx.typing():
        try:
            if attachment:
                if attachment.size > MAX_FILE_SIZE:
                    await ctx.send("❌ File is too large (> 8 MB).")
                    return
                file_bytes = await attachment.read()
            else:
                async with aiohttp.ClientSession() as session:
                    file_bytes = await _download_url(session, url)
        except ValueError as exc:
            await ctx.send(f"❌ {exc}")
            return
        except (aiohttp.ClientError, asyncio.TimeoutError, OSError) as exc:
            await ctx.send(f"❌ Failed to download the file: {exc}")
            return

        # ── 4. Write to temp dir, run dumper, collect output ─────────────────
        stem = _safe_stem(raw_name)
        output_name = f"{stem}.lua.txt"

        with tempfile.TemporaryDirectory() as tmpdir:
            tmp = Path(tmpdir)
            input_path = tmp / f"input{ext}"
            output_path = tmp / output_name

            input_path.write_bytes(file_bytes)

            start = time.monotonic()
            success, error = _run_dumper(input_path, output_path)
            elapsed_ms = int((time.monotonic() - start) * 1000)

            if not success:
                await ctx.send(f"❌ Dumper error: {error}")
                return

            if not output_path.exists() or output_path.stat().st_size == 0:
                await ctx.send("❌ Dumper produced no output.")
                return

            await ctx.send(
                f"✅ file successfully dumped in {elapsed_ms} ms",
                file=discord.File(str(output_path), filename=output_name),
            )


# --------------------------------------------------------------------------- #
# Entry point
# --------------------------------------------------------------------------- #

@bot.event
async def on_ready():
    print(f"Logged in as {bot.user} (id: {bot.user.id})")
    print("Ready. Prefix:", PREFIX)


if __name__ == "__main__":
    if not TOKEN:
        raise RuntimeError(
            "DISCORD_TOKEN environment variable is not set. "
            "Copy .env.example to .env and fill in your token."
        )
    bot.run(TOKEN)
